### PR TITLE
Consolidate data tracked over an async completion session

### PR DESCRIPTION
**Pull request by genlu**
Fix https://github.com/dotnet/roslyn/issues/59262

Based on the public async completion API, it's not mentioned that the same session object would be used over an async completion session, even though the underlying implementation guarantees it. So I have aggregated all data into a single object but still uses the property bag to track it over the session, instead of using a CWT.

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemData.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion
+{
+    internal sealed class CompletionItemData
+    {
+        private const string RoslynCompletionItemData = nameof(RoslynCompletionItemData);
+
+        public RoslynCompletionItem RoslynItem { get; }
+
+        public Optional<SnapshotPoint> TriggerLocation { get; }
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemData.cs:17]: ❔  Can we use `SnapshotPoint?` here?

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemData.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+        public bool IsProvidedByRoslynCompletionSource => TriggerLocation.HasValue;
+
+        private CompletionItemData(RoslynCompletionItem roslynItem, SnapshotPoint? triggerLocation)
+        {
+            RoslynItem = roslynItem;
+            TriggerLocation = triggerLocation ?? new();
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemData.cs:24]: ❕ Don't use `new T()` (or `new()`) for value types unless they have a default constructor. Also, I'm concerned this line may be treating `new()` as `new SnapshotPoint()` and not `new Optional<SnapshotPoint>()` (even if it's not, it's confusing).

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemData.cs (hunk 1 above)
Comment (genlu) [reply src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemData.cs]: Fixed

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemData.cs (hunk 2 above)
Comment (genlu) [reply src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionItemData.cs]: You are right, new() here actually is `new SnapshotPoint()`. Fixed

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSessionData.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    /// </summary>
+    internal sealed class CompletionSessionData
+    {
+        private const string RoslynCompletionSessionData = nameof(RoslynCompletionSessionData);
+        public bool TargetTypeFilterExperimentEnabled { get; set; }
+        public bool TargetTypeFilterSelected { get; set; }
+        public bool HasSuggestionItemOptions { get; set; }
+
+        public Optional<SnapshotPoint> ExpandedItemTriggerLocation { get; set; }
+        public Optional<TextSpan> CompletionListSpan { get; set; }
```
Comment (dibarbet) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSessionData.cs:27]: can all of these be optional at different times?  Or does one being provided mean the rest are provided as well?

**Review by dibarbet — APPROVED** [review]

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs [hunk 4]
```diff
@@ -39,18 +40,11 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncComplet
...
-        internal const string HasSuggestionItemOptions = nameof(HasSuggestionItemOptions);
-        internal const string Description = nameof(Description);
         internal const string PotentialCommitCharacters = nameof(PotentialCommitCharacters);
-        internal const string ExcludedCommitCharacters = nameof(ExcludedCommitCharacters);
         internal const string NonBlockingCompletion = nameof(NonBlockingCompletion);
-        internal const string TargetTypeFilterExperimentEnabled = nameof(TargetTypeFilterExperimentEnabled);
-        internal const string ExpandedItemsTask = nameof(ExpandedItemsTask);
+
+        // Don't change this property! Editor code currently has a dependency on it.
+        internal const string ExcludedCommitCharacters = nameof(ExcludedCommitCharacters);
```
Comment (genlu) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs:47]: OK, turns out editor depend on this property so we have to keep it for now. Only took me one full day of debugging to figure it out...
https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=/src/Editor/Language/Impl/Language/AsyncCompletion/AsyncCompletionSession.cs&version=GBmain&line=1168&lineEnd=1169&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

Comment (genlu) [reply src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs:47]: @olegtk This feels so wrong. Any idea how can we remove this dependency?

**Pull request merged**